vispy/visuals/_scalable_textures.py

```diff
@@ -0,0 +1,441 @@
+        or even require a numpy array to make decisions based on shape **and**
+        dtype.
+    **texture_kwargs
+        Any other keyword arguments to pass to the parent TextureXD class.
+
+    """
+
+    def __init__(self, data=None, **texture_kwargs):
+        self._clim = None
+        self._data_dtype = None
+        data, texture_kwargs = self.init_scaling_texture(data, **texture_kwargs)
+        # Call the __init__ of the TextureXD class
+        super().__init__(data, **texture_kwargs)
+
+    def init_scaling_texture(self, data=None, internalformat=None, **texture_kwargs):
+        """Initialize scaling properties and create a representative array."""
+        self._data_dtype = getattr(data, 'dtype', None)
+        data = self._create_rep_array(data)
+        internalformat = self._get_texture_format_for_data(
+            data,
+            internalformat)
+        texture_kwargs['internalformat'] = internalformat
+        return data, texture_kwargs
+
+    def _get_texture_format_for_data(self, data, internalformat):
+        return internalformat
+
+    @property
+    def clim(self):
+        """Color limits of the texture's data."""
+        return self._clim
+
+    def set_clim(self, clim):
+        """Set clim and return if a texture update is needed.
+
+        In this default implementation, it is assumed changing the color limit
+        never requires re-uploading the data to the texture (always return
+        ``False``).
+
+        """
+        need_texture_upload = False
+        if isinstance(clim, str):
+            if clim != 'auto':
+                raise ValueError('clim must be "auto" if a string')
+            self._clim = clim
+        else:
+            try:
+                cmin, cmax = clim
+            except (ValueError, TypeError):
+                raise ValueError('clim must have two elements')
+            self._clim = (cmin, cmax)
+        return need_texture_upload
+
+    @property
+    def clim_normalized(self):
+        """Normalize current clims to match texture data inside the shader.
+
+        Scaling only happens on the GPU so we only normalize
+        the color limits when needed (for unsigned normalized integer
+        internal formats). Otherwise, for internal formats that are not
+        normalized such as floating point (ex. r32f) we can leave the ``clim``
+        as is.
+
+        """
+        # if the internalformat of the texture is normalized we need to
+        # also normalize the clims so they match in-shader
+        clim_min = self.normalize_value(self.clim[0], self._data_dtype)
+        clim_max = self.normalize_value(self.clim[1], self._data_dtype)
+        return clim_min, clim_max
+
+    @property
+    def is_normalized(self):
+        """Whether the in-shader representation of this texture is normalized or not.
+
+        Formats ending in 'f' (float), 'ui' (unsigned integral), or 'i'
+        (integral) are not normalized in the GPU. Formats ending in "_snorm"
```

Comment:
```suggestion
        Formats ending in 'f' (float), 'ui' (unsigned integer), or 'i'
        (signed integer) are not normalized in the GPU. Formats ending in "_snorm"
```

Reply:
This wording comes from the khronos wiki: https://www.khronos.org/opengl/wiki/Image_Format

I suppose "integer" will make more sense to Python programmers.